lxd/db/openfga/openfga.go

```diff
@@ -28,6 +31,102 @@ func NewOpenFGAStore(clusterDB *db.Cluster) storage.OpenFGADatastore {
+	ReadUsersetTuplesCache    map[string][]*openfgav1.Tuple
+}
+
+// NewOpenFGACache returns a new OpenFGA cache.
+func NewOpenFGACache() *OpenFGACache {
+	return &OpenFGACache{
+		ReadStartingFromUserCache: make(map[string][]*openfgav1.Tuple),
+		ReadUsersetTuplesCache:    make(map[string][]*openfgav1.Tuple),
+	}
+}
+
+// ReadReadStartingFromUserCache reads tuples from the cache.
+func (c *OpenFGACache) ReadReadStartingFromUserCache(filter storage.ReadStartingWithUserFilter) (bool, string, []*openfgav1.Tuple, error) {
+	b, err := json.Marshal(filter)
+	if err != nil {
+		return false, "", nil, err
+	}
+
+	tuples, ok := c.ReadStartingFromUserCache[string(b)]
+	if ok {
+		return true, string(b), tuples, nil
+	}
+
+	return false, string(b), nil, nil
+}
+
+// WriteReadStartingFromUserCache writes tuples to the cache.
+func (c *OpenFGACache) WriteReadStartingFromUserCache(cacheKey string, tuples []*openfgav1.Tuple) {
+	c.ReadStartingFromUserCache[cacheKey] = tuples
+}
+
+// JSONReadUsersetTuplesFilter is an intermediate struct for serializing/deserializing the storage.ReadUsersetTuplesFilter.
+// This is needed because we can't deserialize the storage.ReadUsersetTuplesFilter directly: it contains an interface field that can't be deserialized.
+type JSONReadUsersetTuplesFilter struct {
+	Object                      string `json:"object"`
+	Relation                    string `json:"relation"`
+	AllowedUserTypeRestrictions []struct {
+		Type      string  `json:"type"`
+		Relation  *string `json:"relation"`
+		Wildcard  *bool   `json:"wildcard"`
+		Condition string  `json:"condition"`
+	} `json:"allowed_user_type_restrictions"`
+}
+
+// ReadReadUsersetTuplesCache reads tuples from the cache.
+func (c *OpenFGACache) ReadReadUsersetTuplesCache(filter storage.ReadUsersetTuplesFilter) (bool, string, []*openfgav1.Tuple, error) {
+	jsonFilter := JSONReadUsersetTuplesFilter{
+		Object:   filter.Object,
+		Relation: filter.Relation,
+	}
+
+	if len(filter.AllowedUserTypeRestrictions) > 0 {
+		jsonFilter.AllowedUserTypeRestrictions = make([]struct {
+			Type      string  `json:"type"`
+			Relation  *string `json:"relation"`
+			Wildcard  *bool   `json:"wildcard"`
+			Condition string  `json:"condition"`
+		}, len(filter.AllowedUserTypeRestrictions))
+
+		for i, r := range filter.AllowedUserTypeRestrictions {
+			restriction := &jsonFilter.AllowedUserTypeRestrictions[i]
+			restriction.Type = r.Type
+			restriction.Condition = r.Condition
+
+			switch v := r.RelationOrWildcard.(type) {
+			case *openfgav1.RelationReference_Relation:
+				relation := v.Relation
+				restriction.Relation = &relation
+			case *openfgav1.RelationReference_Wildcard:
+				isWildcard := true
+				restriction.Wildcard = &isWildcard
+			}
+		}
+	}
+
+	b, err := json.Marshal(jsonFilter)
+	if err != nil {
+		return false, "", nil, err
+	}
+
+	tuples, ok := c.ReadUsersetTuplesCache[string(b)]
+	if ok {
+		return true, string(b), tuples, nil
+	}
+
+	return false, string(b), nil, nil
```

Comment:
I'm not sure the JSON marshalling/unmarshalling is necessary for the map keys as I think some fields are ignored by the Datastore. For example, wildcard relations were removed from our model because they introduced more complexity than they were worth for our use-case, so we don't need to handle them.

I would check which fields from the input filters to each function are actually used. Then either concatenate them to get a key, or define a struct containing those fields and use that.

If it turns out that marshalling is absolutely necessary, I *think* we can use `google.golang.org/protobuf/encoding/protojson` to marshal/unmarshal directly, without the intermediate struct.

Reply:
Ok, fixed